addons/mail/models/mail_activity.py

```diff
@@ -795,3 +796,128 @@ def activity_unlink(self, act_type_xmlids, user_id=None):
             domain = ['&'] + domain + [('user_id', '=', user_id)]
         self.env['mail.activity'].search(domain).unlink()
         return True
+
+    def _read_progress_bar(self, domain, group_by, progress_bar):
+        if not (progress_bar['field'] == 'activity_state' and self._fields[group_by].store):
+            return super()._read_progress_bar(domain, group_by, progress_bar)
+
+        # use datetime.utcnow() instead of date.today(), because
+        # it's used in _compute_state_from_date and on mocking in tests
+        today_utc = datetime.utcnow()
+        today = date(year=today_utc.year, month=today_utc.month, day=today_utc.day)
+
+        with_clause = """
+            SELECT
+                mail_activity.res_id AS res_id,
+                min(date_deadline - (%%s AT TIME ZONE res_partner.tz)::timestamp) AS days_to_deadline
+            FROM mail_activity
+            LEFT JOIN res_users ON mail_activity.user_id = res_users.id
+            LEFT JOIN res_partner ON res_users.partner_id = res_partner.id
```

Comment:
I don't understand the comment above about `today`, but it seems these JOINs are only used to compute the `date` in the user's TZ. This is exactly what `fields.Date.context_today(self)` would do, saving you both a bit of code and SQL JOINs that might be expensive, no?

Reply:
``fields.Date.context_today(self)`` computes date for current user, while we need user specified in ``mail.activity``. I don't think that this joins are expensive. Alternative could be adding new field tz to ``mail.activity``, it may make a mess on changing user's timezone.

In the comment I reference to this patch in tests

https://github.com/odoo/odoo/blob/62c849521ef06aeff212cbdd7179a49155276233/addons/test_mail/tests/test_mail_activity.py#L375-L378

Apparently it's in master branch only

Reply:
> `fields.Date.context_today(self)` computes date for current user, while we need user specified in `mail.activity`. 

Though technically you're right, I'm not sure the way we're doing the computation in `_compute_state_from_date`  at the moment is correct (ping @tde-banana-odoo ?)

In Odoo all pure date fields represent a "calendar day" relatively to *the viewer*. They are never converted upon display, only when saving a new value (and that's only to avoid off-by-one effects for timezones with large UTC offsets). E.g an accounting date on an invoice does not change depending on who views it.

Similarly, computing the activity_state based on the responsible's TZ seems incorrect.

Imagine you're a manager and you want to review all the activities that your team is working on *today*. Some of the team members might be in a different timezone, and their workday has not yet started or is already finished - but you still want to look at the activities due "today", i.e. the same calendar day you're on.

Timezone offsets only matter upon display when we want to precisely identify a moment in time. That's not the case here, as far as I can tell.

Note: in practice this won't matter much, because most users work with activities created in the same TZ as they are, so they won't even notice. But it would be too bad to sacrifice performance for the sake of preserving a conversion that is borderline-incorrect in practice.


> I don't think that these joins are expensive.

That's not someone you can just assume, you have to measure it. Once cardinality reaches hundreds of thousands of rows, JOINs are *always* expensive (how could they not?) Here's a comparison of the two versions on a database with 1M partners/users:
- With the JOINs: 292ms https://explain.dalibo.com/plan/15B 
- No JOINs: 80ms https://explain.dalibo.com/plan/vpi

+200ms on an interactive request is *a lot*

Note that the cost is high because the model (`crm.lead`) represents 75% of the mail activities in this case (80k rows), so PG gets no speedup with the `res_model` index. For a model that only take 3% of activities (2.6k rows), the index is used, but the query remains 3-5x faster without the JOINs (e.g. 10ms vs 30ms). That's less dramatic. Except that what we really care about is the performance for large datasets.

Reply:
> Imagine you're a manager and you want to review all the activities that your team is working on today. Some of the team members might be in a different timezone, and their workday has not yet started or is already finished - but you still want to look at the activities due "today", i.e. the same calendar day you're on.

When I thought  about it, I use another example: I am in Europe and some colleagues are in Australia and they already passed midnight. With current implementation I'll see that they have some tasks *overdue*, but if it's computed based on my timezone, I'll have false understanding that everything is fine, while in fact Australian colleagues already sleep and cannot finish the tasks *today*.

Reply:
> * With the JOINs: 292ms https://explain.dalibo.com/plan/15B
> * No JOINs: 80ms https://explain.dalibo.com/plan/vpi
> 
> +200ms on an interactive request is _a lot_

I agree. What I really meant is that adding joins is nothing compare to time spending on original implementation via ORM

Reply:
> When I thought about it, I used another example: I am in Europe and some colleagues are in Australia and they already passed midnight. With current implementation I'll see that they have some tasks _overdue_, but if it's computed based on my timezone, I'd have false understanding that everything is fine, while in fact Australian colleagues already sleep and cannot finish the tasks _today_.

That might be what you could expect with a real-time task scheduling system which would also take into account working hours, holidays, etc. But that's not what we have nor what we want, and assuming your Australian colleagues don't work until midnight, your assumption would be wrong for most of your workday anyway - so that can't be a reliable system.

Reply:
I'd agree with any decision. But do you mean that we can change this behavior in stable branch?